src/types.ts

```diff
@@ -0,0 +1,16 @@
+import { Resolver } from './resolvers'
+
+export interface ResolverInternal<T> extends Resolver<T> {
+  /**
+   * True if this resolver is a value resolver. Used to implicit set the lifetime of a value
+   * resolver to either {@link Lifetime.SCOPED} or {@link Lifetime.SINGLETON} depending on whether
+   * this value is registered to a root or a scope container.
+   */
+  isValue?: boolean
```

Comment:
I think we should make this a bit more generic, maybe `autoLifetime`? I'm trying to avoid coupling these new options to any particular resolver type.

Maybe as an alternative, we could use the root container to resolve any requested `SINGLETON`, that way any overrides would not be used by singletons, but could be used by scoped. Do you think that would have any negative side effects?

If we do this, then we can make `asValue` return `lifetime: SCOPED, isLeakSafe: true` and we'd only need the `isLeakSafe` option.

Reply:
I considered that but changing the resolution behavior would be a breaking change outside of strict mode - unless you're suggesting that strict mode should change the resolution behavior?

If we were to do that, we could avoid leak checking for singletons entirely, since they will always resolve on the root container with only root container elements. Even a scoped registration set on the root container would be safe for a singleton to depend on *because* it will only ever pull that registration from the root container, and not from a potential scope.

Reply:
Actually @jeffijoe I just had an idea that sidesteps this problem entirely. For the purposes of lifetime checking, why not treat a scoped lifetime, defined at the root level, as equivalent to singleton if and only if the resolver used during resolution is that root-registered scoped module (rather than one defined in a scope)?

The only potential gotcha is that the error message for a "real" lifetime issue will be pushed down one level in the resolution stack. e.g. imagine A (singleton) depends on B (scoped, registered at root) depends on C (scoped, registered in scope). In the current setup, A -> B will trigger the error, but in this proposed change, it is only on the resolution of C that this problem is detected.

Reply:
Regarding the first question, I would expect the result to be `foo` because the singleton would be using the registration from the root container; this seems like the most natural to me, as it means it doesn't matter whether it was resolved via a scope or the root, it would always be deterministic.

Now that you mention the case of a singleton depending on a scoped value, I currently don't see a legitimate reason for that ever happening with the exception of the leak-safe resolvers. Perhaps what we should do is have `strict` mode change the resolution behavior to the one we discussed in addition to enforcing these checks. I believe a singleton depending on non-singletons is the big issue we are trying to prevent with this change overall. By doing so, we'd get the (IMO) most natural behavior for `asValue`.

Reply:
Makes sense. I'll update this to the new behavior for strict and push.

Reply:
@jeffijoe it'll be a bit. Changing the target container mid-resolution is a problem - i.e. if you resolve a scoped module, which has a dependency on a singleton, we want to switch resolution to the singleton for strict mode. Currently the assumption is that the container is always the same, and therefore closing over the resolution stack is sufficient since all `resolve` calls from dependencies will target the same container.

As soon as the container is switched due to the singleton in the resolution path, the resolution stack is effectively cleared, meaning that we go one further round before a cyclic dependency is detected and the resolution path indicating a possible error is truncated.

Reply:
Yeah the resolution stack is an implementation detail so shouldn't be exposed. Hmm..

Reply:
@jeffijoe done - I added a new internal container construction function to allow for sharing the same mutable resolution stack array between parents and scopes.

Should we remove the `parentContainer` argument from the public container constructor as well? Technically it's part of the existing public API but it was not documented and should not have been used. If people use that, it will again decouple the resolution stack between parent and child containers and lead to problems in strict mode.

Reply:
Great work @fnimick ! 🙏 

I agree, we should remove `parentContainer`. It was an implementation detail, I don't know why I didn't think of your solution earlier, I've used it plenty of times elsewhere. 😅